tripal_chado/includes/TripalFields/sbo__relationship/sbo__relationship_widget.inc

```diff
@@ -345,97 +325,160 @@ class sbo__relationship_widget extends ChadoFieldWidget {
     $subject_id = isset($form_state['values'][$field_name][$langcode][$delta]['chado-' . $field_table . '__' . $subject_id_key]) ? $form_state['values'][$field_name][$langcode][$delta]['chado-' . $field_table . '__' . $subject_id_key] : '';
     $object_id = isset($form_state['values'][$field_name][$langcode][$delta]['chado-' . $field_table . '__' . $object_id_key]) ? $form_state['values'][$field_name][$langcode][$delta]['chado-' . $field_table . '__' . $object_id_key]: '';
     $type_id = isset($form_state['values'][$field_name][$langcode][$delta]['chado-' . $field_table . '__type_id']) ? $form_state['values'][$field_name][$langcode][$delta]['chado-' . $field_table . '__type_id'] : '';
 
-    $subject_name = isset($form_state['values'][$field_name][$langcode][$delta]['subject_name']) ? $form_state['values'][$field_name][$langcode][$delta]['subject_name']: '';
+    $subject_name = isset($form_state['values'][$field_name][$langcode][$delta]['subject_name']) ? $form_state['values'][$field_name][$langcode][$delta]['subject_name'] : '';
     $object_name = isset($form_state['values'][$field_name][$langcode][$delta]['object_name']) ? $form_state['values'][$field_name][$langcode][$delta]['object_name'] : '';
 
+    // Validation:
+    //------------
     // If the row is empty then skip this one, there's nothing to validate.
-    if (!($type_id or !$type_name) and !$subject_name and !$object_name) {
+    if (!($type_id || $type_name) && !$subject_name && !$object_name) {
+      // @debug print "HERE; Type Name: $type_name, Type ID: $type_id, Subject Name: $subject_name, Object Name: $object_name\n";
       return;
     }
-    else if ($type_name && $voc_id) {
-      $val = array(
-        'cv_id' => $voc_id,
-        'name' => $type_name
-      );
-      $cvterm = chado_generate_var('cvterm', $val);
-      $type_id = $cvterm->cvterm_id;
-    }
 
     // Do not proceed if subject ID or object ID does not exist
     if (!key_exists($subject_id_key, $fkeys[$base_table]['columns']) ||
-      !key_exists($object_id_key, $fkeys[$base_table]['columns'])) {
+        !key_exists($object_id_key, $fkeys[$base_table]['columns'])) {
       return;
     }
-    // Get the subject ID.
-    $subject_id = '';
-    $fkey_rcolumn = $fkeys[$base_table]['columns'][$subject_id_key];
-    $matches = array();
-    if (preg_match('/\[id: (\d+)\]/', $subject_name, $matches)) {
-      $subject_id =  $matches[1];
-    }
-    else {
-      $values = array('uniquename' => $subject_name);
-      $subject = chado_select_record($base_table, array($fkey_rcolumn), $values);
-      if(count($subject) > 0) {
-        $subject_id = $subject[0]->$fkey_rcolumn;
+
+    // Validation is occuring in the field::validate() but we need to know if it finds errors.
+    // As such, I'm calling it here to check.
+    // Also, field::validate() doesn't seem to always show it's errors
```

Comment:
I think this is a broader problem that perhaps deserves its own issue to discuss?

Its very frustrating to "lose" the submission form because the validate didnt prevent submission but there was a field preventing inserting into the db.

Reply:
I think this definitely deserves it's own issue! I simply haven't found time to find a set of steps to demonstrate the issue consistently. Feel free to create the issue if you have!

Reply:
Yeah, this is true of all widgets not just this one.